### PR TITLE
Deny tokens that don't allow approving more than balance

### PR DESCRIPTION
**Pull request by vkgnosis**
Related to https://github.com/gnosis/gp-v2-services/issues/657 . There are tokens that only allow approving current balance which causes us to fail simulations. This PR improves the status quo by deny listing these tokens instead.
A further improvement outlined in the issue is detecting this situation and only approving the balance. I was working on this too but it got more complicated than expected so I wanted to get this PR in first.

### Test Plan
adapted test and checked that a token properly gets deny listed that used to be allowed

### CODE REVIEW

File: shared/src/bad_token/trace_call.rs [hunk 1]
```diff
@@ -369,6 +380,30 @@ mod tests {
...
+                    result: Some(Res::Call(CallResult {
+                        gas_used: 1.into(),
+                        output: Bytes(Vec::new()),
+                    })),
+                    error: None,
+                }]),
+                vm_trace: None,
+                state_diff: None,
+                transaction_hash: None,
+            },
```
Comment (bh2smith) [inline shared/src/bad_token/trace_call.rs:406]: Maybe I am missing something, but which part of this `BlockTrace` indicated that the simulation failed on infinite token approval?

**Review by bh2smith — APPROVED** [review]
Very nice, just one question in the test.

File: shared/src/bad_token/trace_call.rs (hunk 1 above)
Comment (vkgnosis) [reply shared/src/bad_token/trace_call.rs:406]: That test is called handle_response_ok and as such does not test the error case. The error case test is the #ignored mainnet one.

**Review by nlordell — APPROVED** [review]
Nice. In the long run, we may want to allow tokens like this by marking them as "approval limited by balance" so that the generated solution can approve the exact traded amounts. We could even have the orderbook service charge a premium on the fee in those cases to offset the additional cost to the solver.

**Review by fleupold — APPROVED** [review]
This logic could also go in the non-trace based bad token checks, couldn't it?

**Comment by vkgnosis** [comment]
It doesn't need the trace api so it could but I still feel it fits in the bad token detector better.

**Pull request merged**